compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs

```diff
@@ -1233,32 +1228,93 @@ convertExpr env0 e = do
+        else pure $
             ELet (Binding bind' scrutinee') $
-            ECase (EVar $ convVar bind) alts'
+            mkCase env bindTy resultType' (EVar $ convVar bind) alts'
     go env (Let (Rec xs) _) args = unsupported "Local variables defined recursively - recursion can only happen at the top level" $ map fst xs
     go env o@(Coercion _) args = unhandled "Coercion" o
     go _ x args = unhandled "Expression" x
 
+-- | Represents a case alternative with a generalized case pattern.
+data GeneralisedCaseAlternative
+    = GCAEquality EqualityAlternative
+        -- ^ Pattern matching via built-in equality.
+    | GCANormal CaseAlternative
+        -- ^ Normal case alternative that is directly supported by LF.
+        -- This includes the default case (CPDefault).
+    deriving (Eq, Ord)
+
+-- | Is this a normal case alternative?
+isNormalCaseAlternative :: GeneralisedCaseAlternative -> Bool
+isNormalCaseAlternative = \case
+    GCANormal _ -> True
+    _ -> False
+
+-- | Represents a case alternative that is not directly supported
+-- by LF's CaseAlternative syntax. In particular the pattern here
+-- is some expression that we want to use the built-in equality
+-- to match.
+data EqualityAlternative = EqualityAlternative
+    { eqaltPattern :: LF.Expr
+        -- ^ Pattern expression to test against. This is an inert
+        -- expression (can't raise errors, e.g. a constructor).
+    , eqaltBody :: LF.Expr
+        -- ^ Right-hand side of case alternative. Potentially divergent.
+    } deriving (Eq, Ord)
+
+-- | Make a case expression from GeneralisedCaseAlternatives.
+--
+-- The scrutinee will be evaluated multiple times unless all the
+-- case alternatives are normal case alternatives. So to prevent
+-- this, this function should only be used when either the scrutinee
+-- is inert (e.g. a bound variable), or all the alternatives are
+-- normal case alternatives ('GCANormal').
+mkCase :: Env -> LF.Type -> LF.Type -> LF.Expr -> [GeneralisedCaseAlternative] -> LF.Expr
+mkCase env scrutineeType resultType scrutinee galts =
+    finalize (foldr addCaseAlternative (Right []) galts)
+  where
+    finalize :: Either LF.Expr [CaseAlternative] -> LF.Expr
+    finalize (Left e) = e
+    finalize (Right []) = ECase scrutinee
+        [ CaseAlternative CPDefault $ EBuiltin BEError `ETyApp` resultType `ETmApp` EBuiltin (BEText "Unreachable") ]
+        -- GHC only generates empty case alternatives if it is sure the scrutinee will fail.
+        -- LF doesn't support empty alternatives, so we turn this into a non-empty alternative.
+    finalize (Right alts) = ECase scrutinee alts
+
+    addCaseAlternative :: GeneralisedCaseAlternative -> Either LF.Expr [CaseAlternative] -> Either LF.Expr [CaseAlternative]
+    addCaseAlternative (GCANormal alt) (Right alts) = Right (alt : alts)
+    addCaseAlternative (GCANormal alt) (Left e) =
+        case alt of
+            CaseAlternative CPDefault _ -> Right [alt]
```

Comment:
Do we ever hit this? I thought `CPDefault` always comes last where we still have `Right []` in the accumulator.